<a href="https://colab.research.google.com/github/Yuanxue911/psych755/blob/main/assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
pip install transformers datasets


ERROR: Operation cancelled by user
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/pip/_vendor/pkg_resources/__init__.py", line 3108, in _dep_map
    return self.__dep_map
  File "/usr/local/lib/python3.10/dist-packages/pip/_vendor/pkg_resources/__init__.py", line 2901, in __getattr__
    raise AttributeError(attr)
AttributeError: _DistInfoDistribution__dep_map

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/pip/_internal/cli/base_command.py", line 169, in exc_logging_wrapper
    status = run_func(*args)
  File "/usr/local/lib/python3.10/dist-packages/pip/_internal/cli/req_command.py", line 242, in wrapper
    return func(self, options, args)
  File "/usr/local/lib/python3.10/dist-packages/pip/_internal/commands/install.py", line 441, in run
    conflicts = self._determine_conflicts(to_install)
  File "/usr/local/lib/python3.10/dist-packages/pip/_i

In [4]:
from datasets import load_dataset
data_set = load_dataset('fashion_mnist', use_auth_token='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')



/usr/local/lib/python3.10/dist-packages/datasets/load.py:2069: FutureWarning: 'use_auth_token' was deprecated in favor of 'token' in version 2.14.0 and will be removed in 3.0.0.
You can remove this warning by passing 'token=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' instead.
  warnings.warn(


Extracting data files:   0%|          | 0/4 [00:00<?, ?it/s]

Generating train split:   0%|          | 0/60000 [00:00<?, ? examples/s]

Generating test split:   0%|          | 0/10000 [00:00<?, ? examples/s]

In [5]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("bert-base-cased")

In [8]:
print(data_set.keys())



dict_keys(['train', 'test'])


Map:   0%|          | 0/400 [00:00<?, ? examples/s]

In [10]:
import numpy as np
train_data = data_set['train']
test_data = data_set['test']

labels = np.array(train_data["label"])






# Model 1
###Simple dense layer for practice

In [25]:

import numpy as np

train_images = np.array([np.array(image) for image in train_data['image']]) / 255.0
train_labels = np.array(train_data['label'])
test_images = np.array([np.array(image) for image in test_data['image']]) / 255.0
test_labels = np.array(test_data['label'])




In [26]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Flatten, Dense

model = Sequential([
    Flatten(input_shape=(28, 28)),
    Dense(128, activation='relu'),
    Dense(10)
])





In [27]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])


In [28]:
model.fit(train_images, train_labels, epochs=10)


Epoch 1/10
1875/1875 [==============================] - 13s 6ms/step - loss: 0.4989 - accuracy: 0.8252
Epoch 2/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.3740 - accuracy: 0.8654
Epoch 3/10
1875/1875 [==============================] - 12s 6ms/step - loss: 0.3373 - accuracy: 0.8772
Epoch 4/10
1875/1875 [==============================] - 10s 5ms/step - loss: 0.3093 - accuracy: 0.8863
Epoch 5/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.2927 - accuracy: 0.8926
Epoch 6/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.2812 - accuracy: 0.8949
Epoch 7/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.2684 - accuracy: 0.9004
Epoch 8/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.2563 - accuracy: 0.9062
Epoch 9/10
1875/1875 [==============================] - 7s 4ms/step - loss: 0.2459 - accuracy: 0.9068
Epoch 10/10
1875/1875 [==============================] - 9s 5ms/step - loss: 0.

In [29]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
print('\nTest accuracy:', test_acc)



313/313 - 1s - loss: 0.3383 - accuracy: 0.8818 - 985ms/epoch - 3ms/step

Test accuracy: 0.8817999958992004


# Model 2 CNN


### I've create a simple Convolutional Neural Network (CNN) architecture which is different from the simple dense network normally used in model 1

##What the model does:

####The CNN model takes as input a set of 2D images (28x28 pixels). Each image represents an article of clothing, such as a shirt, a shoe, a bag, etc. The goal of the model is to correctly identify the type of clothing present in each image. It does this by learning from the pixel data in the images and the corresponding labels provided during the training phase

In [35]:
# Import necessary modules
import tensorflow as tf
from tensorflow.keras import datasets, layers, models


##How the input/output patterns were chosen:


###The input to this model is an image, represented as a 2D array of pixel intensity values. The output is a probability distribution across the 10 different classes (types of clothing), indicating the model's confidence that the image belongs to each particular class. The model will then predict the class with the highest probability. The input/output pattern is a typical setup for image classification tasks

###Model architecture/structure

In [33]:
# Reshape the images to include a channel dimension (needed for Conv2D layers)
train_images = train_images.reshape(-1, 28, 28, 1)
test_images = test_images.reshape(-1, 28, 28, 1)

# Define the model architecture
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

###In this architecture, the model starts with a series of convolutional and pooling layers, followed by a flattening step, and ends with two dense layers. The convolutional layers are responsible for detecting patterns in the images, while the dense layers interpret these patterns and make a final prediction.

In [34]:
# Add Dense layers on top
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10))

# Compile the model
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

# Train the model
history = model.fit(train_images, train_labels, epochs=10,
                    validation_data=(test_images, test_labels))

# Evaluate the model
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
print('\nTest accuracy:', test_acc)


Epoch 1/10
1875/1875 [==============================] - 86s 45ms/step - loss: 0.4933 - accuracy: 0.8181 - val_loss: 0.3566 - val_accuracy: 0.8715
Epoch 2/10
1875/1875 [==============================] - 78s 42ms/step - loss: 0.3196 - accuracy: 0.8835 - val_loss: 0.3078 - val_accuracy: 0.8883
Epoch 3/10
1875/1875 [==============================] - 77s 41ms/step - loss: 0.2768 - accuracy: 0.8981 - val_loss: 0.2947 - val_accuracy: 0.8971
Epoch 4/10
1875/1875 [==============================] - 77s 41ms/step - loss: 0.2477 - accuracy: 0.9074 - val_loss: 0.2764 - val_accuracy: 0.8987
Epoch 5/10
1875/1875 [==============================] - 79s 42ms/step - loss: 0.2228 - accuracy: 0.9172 - val_loss: 0.2727 - val_accuracy: 0.9010
Epoch 6/10
1875/1875 [==============================] - 76s 41ms/step - loss: 0.2044 - accuracy: 0.9234 - val_loss: 0.2544 - val_accuracy: 0.9062
Epoch 7/10
1875/1875 [==============================] - 78s 42ms/step - loss: 0.1865 - accuracy: 0.9297 - val_loss: 0.2672 -

### Note that shifting from simple dense layer to CNN model the accuracy increased from .88 to.90